jumanji/environments/routing/rware/utils.py

```diff
@@ -0,0 +1,865 @@
+    return grid
+
+
+def place_entities_on_grid(
+    grid: chex.Array, agents: Agent, shelves: Shelf
+) -> chex.Array:
+    """Place agents and shelves on the grid.
+
+    Args:
+        grid: the warehouse floor grid array.
+        agents: a pytree of Agent type containing agent information.
+        shelves: a pytree of Shelf type containing shelf information.
+
+    Returns:
+        the warehouse grid with all agents and shelves placed in their
+        positions.
+    """
+    agent_ids = get_entity_ids(agents)
+    shelf_ids = get_entity_ids(shelves)
+
+    # place agents and shelves on warehouse grid
+    def place_agents_scan(
+        grid_and_agents: Tuple[chex.Array, chex.Array], agent_id: chex.Array
+    ) -> Tuple[Tuple[chex.Array, chex.Array], None]:
+        grid, agents = grid_and_agents
+        grid = place_entity_on_grid(grid, _AGENTS, agents, agent_id)
+        return (grid, agents), None
+
+    def place_shelves_scan(
+        grid_and_shelves: Tuple[chex.Array, chex.Array], shelf_id: chex.Array
+    ) -> Tuple[Tuple[chex.Array, chex.Array], None]:
+        grid, shelves = grid_and_shelves
+        grid = place_entity_on_grid(grid, _SHELVES, shelves, shelf_id)
+        return (grid, shelves), None
+
+    (grid, _), _ = jax.lax.scan(place_agents_scan, (grid, agents), agent_ids)
+    (grid, _), _ = jax.lax.scan(place_shelves_scan, (grid, shelves), shelf_ids)
+    return grid
+
+
+def update_agent(
+    agents: Agent,
+    agent_id: chex.Array,
+    attr: str,
+    value: Union[chex.Array, Position],
+) -> Agent:
+    """Update the attribute information of a specific agent.
+
+    Args:
+        agents: a pytree of either Agent type containing agent information.
+        agent_id: unique ID identifying a specific agent.
+        attr: the attribute to update, e.g. `direction`, or `is_requested`.
+        value: the new value to which the attribute is to be set.
+
+    Returns:
+        the agent with the specified attribute updated to the given value.
+    """
+    params = {attr: value}
+    agent = tree_slice(agents, agent_id)
+    agent = agent._replace(**params)
+    agents: Agent = tree_add_element(agents, agent_id, agent)
+    return agents
+
+
+def update_shelf(
+    shelves: Shelf,
+    shelf_id: chex.Array,
+    attr: str,
+    value: Union[chex.Array, Position],
+) -> Shelf:
+    """Update the attribute information of a specific shelf.
+
+    Args:
+        shelves: a pytree of Shelf type containing shelf information.
+        shelf_id: unique ID identifying a specific shelf.
+        attr: the attribute to update, e.g. `direction`, or `is_requested`.
+        value: the new value to which the attribute is to be set.
+
+    Returns:
+        the shelf with the specified attribute updated to the given value.
+    """
+    params = {attr: value}
+    shelf = tree_slice(shelves, shelf_id)
+    shelf = shelf._replace(**params)
+    shelves: Shelf = tree_add_element(shelves, shelf_id, shelf)
+    return shelves
```

Comment:
This was the original design, but I ran into a very annoying typing issue which I could only solve by splitting them out. Perhaps it would just be simpler to `type: ignore`. But perhaps I can revisit this later.